pkg/snapshot/snapshot.go

```diff
@@ -0,0 +1,335 @@
+package snapshot
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/TykTechnologies/tyk-operator/api/model"
+	tykv1alpha1 "github.com/TykTechnologies/tyk-operator/api/v1alpha1"
+	"github.com/TykTechnologies/tyk-operator/pkg/client/klient"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+)
+
+var (
+	// names stores each ApiDefiniton's ID as a key and .metadata.name field of corresponding ApiDefinition.
+	names map[string]string
+
+	// namesSpaces stores each ApiDefiniton's ID as a key and .metadata.namespace field of corresponding ApiDefinition.
+	nameSpaces map[string]string
+
+	// idAPIs stores ApiDefinitions based on its ID.
+	idAPIs map[string]*model.APIDefinitionSpec
+
+	// ErrNonexistentKey represents an error if the given key does not exist in the object.
+	ErrNonexistentKey = errors.New("key does not exist in the Config Data")
+
+	// ErrNonStringVal represents an error if the underlying value of interface{} is not string type.
+	ErrNonStringVal = errors.New("failed to convert interface{} to string")
+
+	// ErrNonExistentConfigData represents an error if the given ApiDefinition includes empty (nil) ConfigData field.
+	ErrNonExistentConfigData = errors.New("failed to parse ConfigData: non existent")
+)
+
+type Group struct {
+	Policy         tykv1alpha1.SecurityPolicySpec
+	APIDefinitions model.APIDefinitionSpecList
+}
+
+type Groups []Group
+
+const (
+	NameKey      = "k8sName"
+	NamespaceKey = "k8sNamespace"
+	DefaultName  = "REPLACE_ME"
+	DefaultNs    = ""
+)
+
+// PrintSnapshot outputs a snapshot of the Dashboard as a CR.
+func PrintSnapshot(ctx context.Context, apiDefinitionsFile, policiesFile, category string, group bool) error {
+	apiDefSpecList, err := klient.Universal.Api().List(ctx)
+	if err != nil {
+		return err
+	}
+
+	var policiesList []tykv1alpha1.SecurityPolicySpec
+	if policiesFile != "" || group {
+		policiesList, err = klient.Universal.Portal().Policy().All(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	names = make(map[string]string)
+	nameSpaces = make(map[string]string)
+
+	e := json.NewSerializerWithOptions(json.DefaultMetaFactory, nil, nil, json.SerializerOptions{
+		Yaml:   true,
+		Pretty: true,
+		Strict: true,
+	})
+
+	exportApiDefs := func(i int, w io.Writer, v *model.APIDefinitionSpec) error {
+		// Parse Config Data of the ApiDefinition created on Dashboard.
+		name, ns, err := parseConfigData(v, fmt.Sprintf("%s_%d", DefaultName, i))
+		if err != nil {
+			// Instead of terminating the function with error, just do not export ApiDefinitions with invalid ConfigData
+			// fields, instead inform users with the following log message.
+			fmt.Printf("WARNING: failed to parse API %v due to malformed ConfigData, err: %v\n", v.APIID, err)
+			return nil
```

Comment:
Shouldn't it return error here?

Reply:
While exporting ApiDefinition, writing ApiDefinition to a file, if ApiDefinitoin has no valid ConfigData, we do not write that ApiDefinition into the output file; instead, we skip writing its content.

Assume that we have the following ApiDefinitions;
```
ApiDefinition - A (with valid Config Data)
ApiDefinition - B (with invalid Config Data)
ApiDefinition - c (with valid Config Data)
```

The output file will contain ApiDeifnition A and ApiDefinition C.

Reply:
All good on that end. But while processing policy group wise, we skip APIDefinitions if it doesn't have valid configData but according to Carol's comment here 

> When exporting policy, check If a policy reference to an API without k8sName config, output a WARNING message and not include that policy in output.

We should skip processing of that policy all together. But since this function is not returning any error, it will just skip processing of APIs with invalid configData.

Reply:
thank you for spotting this 🙏 i will update the PR accordingly

Reply:
Updated in e4e7fdfcb5b9a9f93839842e7589e68a57951cfe.

Can you please re-review?